test/testsuites/packagestests.go

```diff
@@ -102,7 +107,133 @@ func testOpSrcWithConflictingPackages(t *testing.T) {
+// provided packages.
+func unresolvedMissingSource(t *testing.T) {
+	namespace, err := test.NewTestCtx(t).GetNamespace()
+	assert.NoError(t, err)
+
+	packages := "camel-k-marketplace-e2e-tests-k,missing-package"
+	err = runSourceTest(namespace,
+		"",
+		packages,
+		"Configuring",
+		fmt.Sprintf("Unable to resolve the source - no source contains the requested package(s) [%s]", strings.Replace(packages, ",", " ", -1)),
+	)
+	assert.NoError(t, err)
+}
+
+// nonExistingSource ensures that if the provided source does not exist the CatalogSourceConfig
+// will be placed in the Configuring phase with the expected messages.
+func nonExistingSource(t *testing.T) {
+	namespace, err := test.NewTestCtx(t).GetNamespace()
+	assert.NoError(t, err)
+
+	source := "bad-source"
+	err = runSourceTest(namespace,
+		source,
+		"camel-k-marketplace-e2e-tests",
+		"Configuring",
+		fmt.Sprintf("Provided source (%s) does not exist", source),
+	)
+	assert.NoError(t, err)
+}
+
+// sourceMissingPackages ensures that if the provided source does not contain the expected
+// packages the CatalogSourceConfig will be placed in the Configuring phase with the expected
+// messages.
+func sourceMissingPackages(t *testing.T) {
+	namespace, err := test.NewTestCtx(t).GetNamespace()
+	assert.NoError(t, err)
+
+	err = runSourceTest(namespace,
+		"test-operators",
+		"camel-k-marketplace-e2e-tests,missing-package",
+		"Configuring",
+		"Still resolving package(s) - missing-package. Please make sure these are valid packages within the test-operators OperatorSource.",
+	)
+	assert.NoError(t, err)
+}
+
+func runSourceTest(namespace, source, packages, expectedPhase, expectedMessage string) error {
+	// Get global framework variables.
+	client := test.Global.Client
+
+	// Create a new CatalogSourceConfig with a non-existing targetNamespace.
+	csc := &v2.CatalogSourceConfig{
+		TypeMeta: metav1.TypeMeta{
+			Kind: v2.CatalogSourceConfigKind,
+		}, ObjectMeta: metav1.ObjectMeta{
+			Name:      cscName,
+			Namespace: namespace,
+		},
+		Spec: v2.CatalogSourceConfigSpec{
+			Source:          source,
+			TargetNamespace: "openshift-marketplace",
+			Packages:        packages,
+		}}
+
+	// Create the CatalogSourceConfig and if an error occurs do not run tests that
+	// rely on the existence of the CatalogSourceConfig.
+	// The CatalogSourceConfig is created with nil ctx and must be deleted manually before test suite exits.
+	err := helpers.CreateRuntimeObject(client, nil, csc)
+	if err != nil {
+		return err
+	}
+
+	// Check that the CatalogSourceConfig with an non-existing targetNamespace eventually reaches the
+	// configuring phase with the expected message.
+	err = helpers.WaitForExpectedPhaseAndMessage(test.Global.Client, cscName, namespace, expectedPhase, expectedMessage)
+	if err != nil {
+		return err
+	}
+	// Delete the CatalogSourceConfig.
+	err = helpers.DeleteRuntimeObject(client, csc)
+	if err != nil {
+		return err
+	}
+	helpers.WaitForNotFound(client, csc, namespace, cscName)
```

Comment:
You need to return this but this could result in an error

Reply:
> Have you done manual testing that our upgrade flow continues to work? i.e. updates in Quay makes it way down to a cluster. This does not mean that you simulated the scenario without actual manual e2e testing.

I performed the following manual E2E test:
# Manual Sync Testing

## Testing OperatorSource Syncs

I started by updating the sync time:
```go
const (
	resyncInterval = time.Duration(2) * time.Minute

	initialWait                = time.Duration(1) * time.Minute
	updateNotificationSendWait = time.Duration(1) * time.Minute
)
```

I then created the following OpSrc, this AppRegistry only contains twistlock:
```yaml
apiVersion: "operators.coreos.com/v1"
kind: OperatorSource
metadata:
  name: agreene-operators
  namespace: openshift-marketplace
spec:
  type: appregistry
  endpoint: https://quay.io/cnr
  registryNamespace: agreene
  displayName: "Agreene Operators"
  publisher: "Agreene"
```

Created the following CSC, it should complain that the `agreene-operators opsrc` is missing the syndesis package:
```yaml
apiVersion: "operators.coreos.com/v1"
kind: "CatalogSourceConfig"
metadata:
  name: "agreene-syndesis-twistlock"
  namespace: "openshift-marketplace"
spec:
  source: agreene-operators
  targetNamespace: openshift-operators
  packages: syndesis,twistlock
```

```bash
$ oc get csc
NAME                         STATUS        MESSAGE                                                                                                                         AGE
agreene-syndesis-twistlock   Configuring   Still resolving package(s) - syndesis. Please make sure these are valid packages within the agreene-operators OperatorSource.   25s
certified-operators          Succeeded     The object has been successfully reconciled                                                                                     27m
community-operators          Succeeded     The object has been successfully reconciled                                                                                     27m
redhat-operators             Succeeded     The object has been successfully reconciled                                                                                     27m
```

Updated the agreene-operators AppRegistry to include the syndesis operator.
```bash
$ operator-courier --verbose push syndesis agreene syndesis 0.0.1 "$AUTH_TOKEN"
```

I made the syndesis the operator public and waited 2 minutes for the OperatorSource sync to kickoff...

After the OperatorSource was synced, the CSC was reconciled successfully:
```bash
$ oc get csc
NAME                         STATUS      MESSAGE                                       AGE
agreene-syndesis-twistlock   Succeeded   The object has been successfully reconciled   86s
certified-operators          Succeeded   The object has been successfully reconciled   28m
community-operators          Succeeded   The object has been successfully reconciled   28m
redhat-operators             Succeeded   The object has been successfully reconciled   28m
```
NOTE: Only `agreene-syndesis-twistlock CSC was updated. 

The CSC deployment created:
```bash
$ oc get pods
NAME                                          READY   STATUS    RESTARTS   AGE
agreene-operators-596867797-sz985             1/1     Running   0          93s
agreene-syndesis-twistlock-678f4744f7-49ffd   1/1     Running   0          95s
certified-operators-86cb6b899d-8fnf5          1/1     Running   0          4m28s
community-operators-f75cb757-m7v42            1/1     Running   0          4m25s
redhat-operators-6d8cdb75f7-w4xq8             1/1     Running   0          4m23s
```

I then updated the syndesis version:
```bash
$ operator-courier --verbose push syndesis agreene syndesis 0.0.2 "$AUTH_TOKEN"
```

After the OperatorSource Sync occured the `agreene-syndesis-twistlock` CSC updated with new version:
```bash
$ oc get csc agreene-syndesis-twistlock -o yaml
apiVersion: operators.coreos.com/v2
kind: CatalogSourceConfig
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"operators.coreos.com/v1","kind":"CatalogSourceConfig","metadata":{"annotations":{},"name":"agreene-syndesis-twistlock","namespace":"openshift-marketplace"},"spec":{"packages":"syndesis,twistlock","source":"agreene-operators","targetNamespace":"openshift-operators"}}
  creationTimestamp: "2019-06-20T13:44:35Z"
  finalizers:
  - finalizer.catalogsourceconfigs.operators.coreos.com
  generation: 11
  name: agreene-syndesis-twistlock
  namespace: openshift-marketplace
  resourceVersion: "21777"
  selfLink: /apis/operators.coreos.com/v2/namespaces/openshift-marketplace/catalogsourceconfigs/agreene-syndesis-twistlock
  uid: 8aa83e31-9361-11e9-b656-0af42a45c8a4
spec:
  csDisplayName: Custom
  csPublisher: Custom
  packages: syndesis,twistlock
  source: agreene-operators
  targetNamespace: openshift-operators
status:
  currentPhase:
    lastTransitionTime: "2019-06-20T13:49:42Z"
    lastUpdateTime: "2019-06-20T13:49:42Z"
    phase:
      message: The object has been successfully reconciled
      name: Succeeded
  packageRepositioryVersions:
    syndesis: 0.0.2
    twistlock: 0.0.1
```

I then removed the syndesis operator from the agreene appRegistry. After the OperatorSource was synced the `agreene-syndesis-twistlock` CSC updated to not include syndesis:

```bash
$ oc get csc agreene-syndesis-twistlock -o yaml
apiVersion: operators.coreos.com/v2
kind: CatalogSourceConfig
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"operators.coreos.com/v1","kind":"CatalogSourceConfig","metadata":{"annotations":{},"name":"agreene-syndesis-twistlock","namespace":"openshift-marketplace"},"spec":{"packages":"syndesis,twistlock","source":"agreene-operators","targetNamespace":"openshift-operators"}}
  creationTimestamp: "2019-06-20T13:44:35Z"
  finalizers:
  - finalizer.catalogsourceconfigs.operators.coreos.com
  generation: 21
  name: agreene-syndesis-twistlock
  namespace: openshift-marketplace
  resourceVersion: "25533"
  selfLink: /apis/operators.coreos.com/v2/namespaces/openshift-marketplace/catalogsourceconfigs/agreene-syndesis-twistlock
  uid: 8aa83e31-9361-11e9-b656-0af42a45c8a4
spec:
  csDisplayName: Custom
  csPublisher: Custom
  packages: twistlock
  source: agreene-operators
  targetNamespace: openshift-operators
status:
  currentPhase:
    lastTransitionTime: "2019-06-20T14:00:19Z"
    lastUpdateTime: "2019-06-20T14:00:19Z"
    phase:
      message: The object has been successfully reconciled
      name: Succeeded
  packageRepositioryVersions:
    twistlock: 0.0.1
```